### PR TITLE
try to calculate width of UTF-8 encoded characters

### LINKED ISSUE #12014: Progress bar width calculation does not recognize double-width characters
## Describe the bug

When the progress bar is rendered wider than expected, an unexpected line break occurs, causing scrolling, and a failure to clear the bar whose content is on the second to last line for the most part.

`unicode.nix`
```
with import <nixpkgs> { };
runCommand
"unicod"
{ }
''
  echo 1🔍4---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  sleep 1
  echo a零b---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  sleep 1
  echo ok
  sleep 1
  touch $out
''
```

```console
$ nix build -f unicode.nix
[1/0/1 built, 0.0 MiB DL] building unicod on ssh://agent-2: 1🔍4----------------------------------------------------------------------------
[1/0/1 built, 0.0 MiB DL] building unicod on ssh://agent-2: a零b----------------------------------------------------------------------------

$
```

Note `-L` was not passed, so it is not expected for log lines to remain on screen, and indeed the `ok` line did not.

## Steps To Reproduce

1. Write a build that prints some `🔍` to the log and sleeps to make sure the line makes it into the progress bar.
1. Build it.
1. Observe a unintentional snapshot of the log on the second-to-last line of the terminal

## Expected behavior

Double-width unicode characters count double when deciding where to cut off the progress bar content.

In Rust, I would use [`unicode_width`](https://docs.rs/unicode-width/0.2.0/unicode_width/). I am not aware of a C/C++ alternative.

Assuming this logic will come from a dependency, we'll want to make it optional, in order to keep the compile-from-scratch footprint small.

I personally wouldn't mind introducing optional Rust code into the project, but that would have to be a team decision.

## Additional context

This affects non-western users and users of tools with fancy logging.

Loosely related:
- #7923

## Checklist

[latest Nix manual]: https://nixos.org/manual/nix/unstable/
[source]: https://github.com/NixOS/nix/tree/master/doc/manual/source
[open bug issues and pull requests]: https://github.com/NixOS/nix/labels/bug

---

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by NaN-git**
## Motivation

Nix assumes that each UTF-8 encoded character has a maximum width of `1`, which can corrupt the terminal output, when wider characters are printed.

An additional header-only dependency `widechar_width.h` is introduced, which is basically a set of large lookup tables. I'm not sure whether the addition of the dependency is acceptable. Personally, I would have copied the lookup tables into the Nix source code.

## Context

- Fixes #12014, includes tests

The functionality can be extended to verify UTF-8 encoded strings and to convert them to UCS-4 encoded strings.

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libutil/package.nix [hunk 1]
```diff
@@ -17,6 +17,19 @@
...
   inherit (lib) fileset;
+  widecharwidth = stdenv.mkDerivation {
+    name = "widecharwidth";
+    dontUnpack = true;
+    dontStrip = true;
+    installPhase = ''
+      mkdir -p $out/include
+      cp $src $out/include/widechar_width.h
+    '';
+    src = builtins.fetchurl {
```
Comment (Mic92) [inline src/libutil/package.nix:28]: I would prefer if we clone the whole repository. Also we shouldn't use builtins fetcher.

```suggestion
    src = pkgs.fetchFromGitHub {
```

File: src/libutil/package.nix [hunk 2]
```diff
@@ -38,6 +51,10 @@ mkMesonLibrary (finalAttrs: {
     (fileset.fileFilter (file: file.hasExt "hh") ./.)
   ];
 
+  nativeBuildInputs = [
+    widecharwidth
```
Comment (Mic92) [inline src/libutil/package.nix:55]: Should be native build inputs, no?

File: src/libutil/package.nix [hunk 3]
```diff
@@ -65,6 +82,7 @@ mkMesonLibrary (finalAttrs: {
   mesonFlags = [
     (lib.mesonEnable "cpuid" stdenv.hostPlatform.isx86_64)
   ];
+  NIX_CFLAGS_COMPILE = "-I${widecharwidth}/include";
```
Comment (Mic92) [inline src/libutil/package.nix:85]: Than we also shouldn't need this.

Comment (roberth) [reply src/libutil/package.nix:85]: Probably needs a `pkg-config` file then.

Ideally we could add that upstream.
Also this should be packaged in Nixpkgs, but developing the expression here is quite convenient.

Comment (NaN-git) [reply src/libutil/package.nix]: I updated the code. Would the current definition of the derivation be ready for nixpkgs?

File: src/libutil/terminal.cc [hunk 4]
```diff
@@ -11,6 +11,53 @@
...
+    for (size_t i = 1; i < bytes; i++) {
+        if (i < s.size() && (s[i] & 0xc0) == 0x80) {
+            ch = (ch << 6) | (s[i] & 0x3f);
+        } else {
+            return {i, i}; // invalid UTF-8 encoding; assume one character per byte
+        }
+    }
+    int width = bytes; // in case of overlong encoding
+    if (ch < max) {
+        width = widechar_wcwidth(ch);
```
Comment (Mic92) [inline src/libutil/terminal.cc:48]: Can you think of a test, we could add here?

Comment (NaN-git) [reply src/libutil/terminal.cc:48]: I added tests for `filterANSIEscapes`

**Pull request merged**